After stop-checks, if daemon is restarted it panics with "change X unexpectedly became unready"

I got Pebble into a funky state today, where it's panicking on startup with the error "change X unexpectedly became unready (Abort)". I've figured out how to repro this now:

1. Configure Pebble with a single health check:

```
$ cat /var/lib/pebble/default/layers/001-base.yaml 
checks:
  chk1:
    override: replace
    exec:
      command: echo foo
```

2. Ensure there's no previous state file: `rm /var/lib/pebble/default/.pebble.state`
3. Start the Pebble daemon: `PEBBLE_DEBUG=1 go run ./cmd/pebble run`
4. In another terminal, stop the check: `go run ./cmd/pebble stop-checks chk1`
5. Back in the `pebble run` terminal, press Ctrl-C to exit the daemon
6. Restart the daemon: `PEBBLE_DEBUG=1 go run ./cmd/pebble run`
7. Watch it panic

Full log of the `pebble run` terminal here:

```
$ rm /var/lib/pebble/default/.pebble.state 

$ PEBBLE_DEBUG=1 go run ./cmd/pebble run
2025-05-28T00:26:33.299Z [pebble] DEBUG Reaper started, waiting for SIGCHLD.
2025-05-28T00:26:33.313Z [pebble] DEBUG socket "/var/lib/pebble/default/.pebble.socket" was not activated; listening
2025-05-28T00:26:33.313Z [pebble] Started daemon.
2025-05-28T00:26:33.315Z [pebble] DEBUG activation done in 16ms
2025-05-28T00:26:33.315Z [pebble] DEBUG Running task 1 on Do: Perform exec check "chk1"
2025-05-28T00:26:33.318Z [pebble] DEBUG Performing check "chk1" with period 10s
2025-05-28T00:26:33.318Z [pebble] DEBUG pid=909318;uid=1000;socket=/var/lib/pebble/default/.pebble.socket; POST /v1/services 2.488601ms 400 (http+unix)
2025-05-28T00:26:33.318Z [pebble] POST /v1/services 2.488601ms 400 (http+unix)
2025-05-28T00:26:33.319Z [pebble] Cannot start default services: no default services
2025-05-28T00:26:35.565Z [pebble] DEBUG pid=909462;uid=1000;socket=/var/lib/pebble/default/.pebble.socket; POST /v1/checks 4.365266ms 200 (http+unix)
2025-05-28T00:26:35.565Z [pebble] POST /v1/checks 4.365266ms 200 (http+unix)
2025-05-28T00:26:35.565Z [pebble] DEBUG Check "chk1" stopped during perform-check (no error)
^C2025-05-28T00:26:37.446Z [pebble] Exiting on interrupt signal.
2025-05-28T00:26:37.446Z [pebble] DEBUG No services to stop.
2025-05-28T00:26:37.447Z [pebble] DEBUG State engine is stopping *state.TaskRunner
2025-05-28T00:26:37.447Z [pebble] DEBUG State engine is stopping *logstate.LogManager
2025-05-28T00:26:37.447Z [pebble] DEBUG Reaper stopped.

$ PEBBLE_DEBUG=1 go run ./cmd/pebble run
2025-05-28T00:26:38.426Z [pebble] DEBUG Reaper started, waiting for SIGCHLD.
2025-05-28T00:26:38.438Z [pebble] DEBUG socket "/var/lib/pebble/default/.pebble.socket" was not activated; listening
2025-05-28T00:26:38.439Z [pebble] Started daemon.
2025-05-28T00:26:38.439Z [pebble] DEBUG activation done in 13ms
panic: change 1 unexpectedly became unready (Abort)

goroutine 21 [running]:
github.com/canonical/pebble/internals/overlord/state.(*Change).detectChangeReady(0xc0002b40b0, 0xc0002b6500)
	/home/ben/w/pebble/internals/overlord/state/change.go:501 +0x185
github.com/canonical/pebble/internals/overlord/state.(*Change).taskStatusChanged(0xc0002b40b0, 0xc0002b6500, 0x6, 0x4)
	/home/ben/w/pebble/internals/overlord/state/change.go:517 +0xa5
github.com/canonical/pebble/internals/overlord/state.(*Task).changeStatus(0xc0002b6500, 0x6, 0x4)
	/home/ben/w/pebble/internals/overlord/state/task.go:250 +0xf9
github.com/canonical/pebble/internals/overlord/state.(*Task).SetStatus(0xc0002cc070?, 0xc0002b6500?)
	/home/ben/w/pebble/internals/overlord/state/task.go:271 +0x47
github.com/canonical/pebble/internals/overlord/state.(*TaskRunner).Ensure(0xc0002cc070)
	/home/ben/w/pebble/internals/overlord/state/taskrunner.go:446 +0x8b7
github.com/canonical/pebble/internals/overlord.(*StateEngine).Ensure(0xc0002a0410?)
	/home/ben/w/pebble/internals/overlord/stateengine.go:141 +0x196
github.com/canonical/pebble/internals/overlord.(*Overlord).Loop.func1()
	/home/ben/w/pebble/internals/overlord/overlord.go:416 +0x3d
gopkg.in/tomb%2ev2.(*Tomb).run(0xc0002b0000, 0x0?)
	/home/ben/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x2b
created by gopkg.in/tomb%2ev2.(*Tomb).Go in goroutine 1
	/home/ben/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xdb
exit status 2
```

--- a/internals/overlord/checkstate/manager.go
+++ b/internals/overlord/checkstate/manager.go
@@ -600,7 +600,6 @@ func (m *CheckManager) StopChecks(checks []string) (stopped []string, err error)
 		change := m.state.Change(checkData.changeID)
 		if change != nil {
 			change.Abort()
-			change.SetStatus(state.AbortStatus)
 			stopped = append(stopped, check.Name)
 		}
 		// We pass in the current number of failures so that it remains the

--- a/internals/overlord/checkstate/manager_test.go
+++ b/internals/overlord/checkstate/manager_test.go
@@ -737,39 +737,80 @@ func (s *ManagerSuite) TestStopChecks(c *C) {
 	}
 	err := s.planMgr.AppendLayer(origLayer, false)
 	c.Assert(err, IsNil)
+
+	// Run an Ensure pass to kick the check tasks into Doing status.
+	st := s.overlord.State()
+	st.EnsureBefore(0)
+
 	waitChecks(c, s.manager, []*checkstate.CheckInfo{
 		{Name: "chk1", Startup: "enabled", Status: "up", Threshold: 3},
 		{Name: "chk2", Startup: "disabled", Status: "inactive", Threshold: 3},
 		{Name: "chk3", Startup: "enabled", Status: "up", Threshold: 3},
 	})
+
 	checks, err := s.manager.Checks()
 	c.Assert(err, IsNil)
-	var originalChangeIDs []string
+	var chk1ChangeID, chk3ChangeID string
 	for _, check := range checks {
-		originalChangeIDs = append(originalChangeIDs, check.ChangeID)
+		switch check.Name {
+		case "chk1":
+			chk1ChangeID = check.ChangeID
+		case "chk3":
+			chk3ChangeID = check.ChangeID
+		}
+	}
+	c.Assert(chk1ChangeID, Not(Equals), "")
+	c.Assert(chk3ChangeID, Not(Equals), "")
+
+	start := time.Now()
+	for {
+		if time.Since(start) > 3*time.Second {
+			c.Fatal("timed out waiting for check changes to go to Doing")
+		}
+		st.Lock()
+		chk1Status := st.Change(chk1ChangeID).Status()
+		chk3Status := st.Change(chk3ChangeID).Status()
+		st.Unlock()
+		if chk1Status == state.DoingStatus && chk3Status == state.DoingStatus {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
 	}
 
 	changed, err := s.manager.StopChecks([]string{"chk1", "chk2"})
+	c.Assert(err, IsNil)
+	c.Assert(changed, DeepEquals, []string{"chk1"})
+
+	// Run an Ensure pass to actually stop the checks.
+	st.EnsureBefore(0)
+
 	waitChecks(c, s.manager, []*checkstate.CheckInfo{
 		{Name: "chk1", Startup: "enabled", Status: "inactive", Threshold: 3},
 		{Name: "chk2", Startup: "disabled", Status: "inactive", Threshold: 3},
 		{Name: "chk3", Startup: "enabled", Status: "up", Threshold: 3},
 	})
-	c.Assert(err, IsNil)
-	c.Assert(changed, DeepEquals, []string{"chk1"})
+
+	// chk3 should still have the same change ID, chk1 and chk2 should not have one.
 	checks, err = s.manager.Checks()
 	c.Assert(err, IsNil)
-	// chk3 should still have the same change ID, chk1 and chk2 should not have one.
 	c.Assert(checks[0].ChangeID, Equals, "")
 	c.Assert(checks[1].ChangeID, Equals, "")
-	c.Assert(checks[2].ChangeID, Equals, originalChangeIDs[2])
-	// chk1's old Change should have aborted.
-	st := s.overlord.State()
-	st.Lock()
-	change := st.Change(originalChangeIDs[0])
-	status := change.Status()
-	st.Unlock()
-	c.Assert(status, Equals, state.AbortStatus)
+	c.Assert(checks[2].ChangeID, Equals, chk3ChangeID)
+
+	// chk1's old change should go to Done.
+	start = time.Now()
+	for {
+		if time.Since(start) > 3*time.Second {
+			c.Fatal("timed out waiting for check changes to go to Done")
+		}
+		st.Lock()
+		chk1Status := st.Change(chk1ChangeID).Status()
+		st.Unlock()
+		if chk1Status == state.DoneStatus {
+			break
+		}
+		time.Sleep(10 * time.Millisecond)
+	}
 }
 
 func (s *ManagerSuite) TestStopChecksNotFound(c *C) {

